tests/integration/SPIKE.md

```diff
@@ -0,0 +1,156 @@
+<h4> Report on testing CLI with EvalAI server on Production/Staging branch on CI/CD pipeline </h4>
+
+Testing is a crucial part of the CI/CD process. Unit tests cover individual units in the code by using a stubbing
+approach i.e. pre-defined responses that will be used for a specific request. However, this does not give a complete
+picture of how all the components interact.
+
+![Stubbing Image](../../images/stubbing.png)
+
+Currently on the CI/CD pipeline for the EvalAI-CLI project, only unit tests are run. In the actual application, we use
+the CLI in integration with the EvalAI server. To test this complete setup, integration tests are required.
+
+The following is a SPIKE report on implementing integration tests to check the interaction of CLI with the server.
+
+<h5> Environment  Setup </h5>
+In addition to the basic setup for the CLI, the server environment needs to be set up.
+
+For the environment setup, the alternatives are:
+
+<h6> Run EvalAI server in development environment in Travis VM </h6>
+
+The server should be set up on Travis-CI VM so as to:
+  1.	Ensure that the tests run fast.
+  2.	Avoid any errors/build failures caused due to issues in the evalapi server.
+  3.	Avoid blocking of CI process when for example, the evalapi server is down.
+
+For this setup to work, the EvalAI server needs to be set up as follows:
+  1.	Creating token setup for users required for the test (e.g. participant1, participant2, host1, host2, etc.)
+  2.	Setting up the database that the server will be modifying: as our objective is only to test the interaction
+  between the CLI and the server, this part can be mocked.
+
+  ![Mock database](../../images/mock-db.png)
+
+However, setting up the environment this way can take a lot of time. On the Travis VM, setting up the server took about
+8-10 minutes on average. This makes the method unsuitable for regular testing on CI/CD pipeline. But, this can be
+implemented successfully on the Production/Staging branches where updates are less frequent. For example, the build can
```

Comment:
@nikochiko 
Can you try if https://docs.travis-ci.com/user/build-stages/ is feasible as we can try to run unit test and integration test in parallel. Using this, we can get feedback from Unit Test and _maybe, integration test will only run on some branches._

Reply:
Aha. Right. I think that's a good choice. While the unit tests take only 35-45 seconds, and the integration tests take 8-10 minutes, it wouldn't make much difference even if they were run at the same time paralelly. But we can set the integration tests to only run after the unit tests are passing. This way, a lot of time will be saved. And integration tests can be implemented on more branches. I hope this is what you mean... !? Please correct me if I'm wrong 😅. I will add this in the report 😄